File: backend/src/test/java/lan/dk/podcastserver/manager/downloader/FfmpegDownloaderTest.java
package lan.dk.podcastserver.manager.downloader;


import com.github.davinkevin.podcastserver.service.M3U8Service;
import com.github.davinkevin.podcastserver.service.MimeTypeService;
import com.github.davinkevin.podcastserver.service.ProcessService;
import com.github.davinkevin.podcastserver.service.UrlService;
import io.vavr.control.Try;
import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.entity.Status;
import lan.dk.podcastserver.manager.ItemDownloadManager;
import lan.dk.podcastserver.repository.ItemRepository;
import lan.dk.podcastserver.repository.PodcastRepository;
import com.github.davinkevin.podcastserver.service.FfmpegService;
import lan.dk.podcastserver.service.properties.PodcastServerParameters;
import com.github.davinkevin.podcastserver.IOUtils;
import net.bramp.ffmpeg.builder.FFmpegBuilder;
import net.bramp.ffmpeg.progress.Progress;
import net.bramp.ffmpeg.progress.ProgressListener;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.util.FileSystemUtils;

import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

import static com.jayway.awaitility.Awaitility.await;
import static io.vavr.API.List;
import static io.vavr.API.Try;
import static java.util.concurrent.CompletableFuture.runAsync;
import static lan.dk.podcastserver.assertion.Assertions.assertThat;
import static com.github.davinkevin.podcastserver.IOUtils.ROOT_TEST_PATH;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyString;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.anyVararg;
import static org.mockito.Mockito.*;

/**
 * Created by kevin on 20/02/2016 for Podcast Server
 */
@RunWith(MockitoJUnitRunner.class)
public class FfmpegDownloaderTest {

    private @Mock PodcastRepository podcastRepository;
    private @Mock ItemRepository itemRepository;
    private @Mock PodcastServerParameters podcastServerParameters;
    private @Mock SimpMessagingTemplate template;
    private @Mock MimeTypeService mimeTypeService;
    private @Mock ItemDownloadManager itemDownloadManager;

    private @Mock UrlService urlService;
    private @Mock M3U8Service m3U8Service;
    private @Mock FfmpegService ffmpegService;
    private @Mock ProcessService processService;

    private @InjectMocks
    FfmpegDownloader ffmpegDownloader;

    Podcast podcast;
    Item item;

    @Before
    public void beforeEach() {
        podcast = Podcast.builder()
                .title("M3U8Podcast")
                .build();

        item = Item.builder()
                .podcast(podcast)
                .url("http://foo.bar/com.m3u8")
                .status(Status.STARTED)
                .numberOfFail(0)
                .build();

        ffmpegDownloader.setItemDownloadManager(itemDownloadManager);
        ffmpegDownloader.setDownloadingItem(DownloadingItem.builder().item(item).urls(List(item.getUrl())).userAgent("Fake UserAgent").build());
        when(podcastServerParameters.getRootfolder()).thenReturn(IOUtils.ROOT_TEST_PATH);
        when(podcastServerParameters.getDownloadExtension()).thenReturn(".psdownload");
        when(podcastRepository.findById(eq(podcast.getId()))).thenReturn(Optional.of(podcast));
        when(itemRepository.save(any(Item.class))).then(i -> i.getArguments()[0]);

        FileSystemUtils.deleteRecursively(ROOT_TEST_PATH.resolve(podcast.getTitle()).toFile());
        Try(() -> Files.createDirectories(ROOT_TEST_PATH));
        ffmpegDownloader.postConstruct();
    }

    @Test
    public void should_download_file() {
        /* Given */
        when(ffmpegService.getDurationOf(anyString(), anyString())).thenReturn(1_000_000D);
        when(ffmpegService.download(anyString(), any(FFmpegBuilder.class), any(ProgressListener.class))).then(i -> {
            FFmpegBuilder builder = i.getArgument(1);
            String location = builder.build().stream().filter(s -> s.contains(ROOT_TEST_PATH.resolve(podcast.getTitle()).toString())).findFirst().orElseThrow(RuntimeException::new);
            Files.write(Paths.get(location), "".getBytes(), StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING );
            Progress progress = new Progress();
            progress.out_time_ms = 90_000;
            ((ProgressListener) i.getArgument(2)).progress(progress);
            return mock(Process.class);
        });
        when(processService.waitFor(any())).thenReturn(new arrow.core.Try.Success<>(1));
        doAnswer(i -> {
            Path targetLocation = i.getArgument(0);
            Files.write(targetLocation, "".getBytes(), StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
            return null;
        }).when(ffmpegService).concat(any(Path.class), anyVararg());

        /* When */
        Item downloaded = ffmpegDownloader.download();

        /* Then */
        assertThat(IOUtils.ROOT_TEST_PATH.resolve(podcast.getTitle()).resolve(item.getFileName())).exists();
        assertThat(downloaded).isSameAs(item);
        assertThat(downloaded.getStatus()).isSameAs(Status.FINISH);
        assertThat(item).hasProgression(9);
    }

    @Test
    public void should_be_compatible() {
        assertThat(ffmpegDownloader.compatibility(DownloadingItem.builder().urls(List(item.getUrl())).build())).isLessThan(Integer.MAX_VALUE/2);
        assertThat(ffmpegDownloader.compatibility(DownloadingItem.builder().urls(List("http://foo.bar/things.rss")).build())).isEqualTo(Integer.MAX_VALUE);
    }

    @Test
    public void should_not_be_compatible() {
        assertThat(ffmpegDownloader.compatibility(DownloadingItem.builder().urls(List("http://foo.bar/things.rss")).build())).isEqualTo(Integer.MAX_VALUE);
    }

    @Test
    public void should_restart_a_current_download() throws InterruptedException {
        /* Given */
        Process downloadProcess = mock(Process.class);
        item.setStatus(Status.PAUSED);
        when(ffmpegService.download(anyString(), any(), any())).thenReturn(downloadProcess);
        when(processService.start(any())).thenReturn(mock(Process.class));

        /* When */
        runAsync(() -> ffmpegDownloader.download());
        Try.run(() -> TimeUnit.MILLISECONDS.sleep(50));
        ffmpegDownloader.restartDownload();

        /* Then */
        await().atMost(5, TimeUnit.SECONDS).until(() -> {
            assertThat(item).hasStatus(Status.STARTED);
        });
    }

    @Test
    public void should_failed_to_restart() {
        /* Given */
        Process downloadProcess = mock(Process.class);
        item.setStatus(Status.PAUSED);
        when(ffmpegService.download(anyString(), any(), any())).thenReturn(downloadProcess);
        doThrow(RuntimeException.class).when(processService).start(any());

        /* When */
        runAsync(() -> ffmpegDownloader.download());
        Try.run(() -> TimeUnit.MILLISECONDS.sleep(50));
        ffmpegDownloader.restartDownload();

        /* Then */
        await().atMost(5, TimeUnit.SECONDS).until(() -> {
            assertThat(item).hasStatus(Status.FAILED);
        });
    }

    @Test
    public void should_paused_a_download() {
        /* Given */
        Process process = mock(Process.class);
        when(ffmpegService.download(anyString(), any(), any())).thenReturn(process);
        when(processService.waitFor(any())).then(i -> {
            TimeUnit.SECONDS.sleep(20L);
            return new arrow.core.Try.Success<>(10);
        });
        when(processService.start(any())).thenReturn(mock(Process.class));

        /* When */
        CompletableFuture<Void> future = runAsync(() -> ffmpegDownloader.download());
        Try.run(() -> TimeUnit.MILLISECONDS.sleep(50));
        ffmpegDownloader.pauseDownload();

        /* Then */
        await().atMost(2, TimeUnit.SECONDS).until(() -> {
            assertThat(item).hasStatus(Status.PAUSED);
        });
        future.cancel(true);
    }
    
    @Test
    public void should_failed_to_pause() {
        /* Given */
        Process process = mock(Process.class);
        when(ffmpegService.download(anyString(), any(), any())).thenReturn(process);
        when(processService.waitFor(any())).then(i -> {
            TimeUnit.SECONDS.sleep(20L);
            return new arrow.core.Try.Success<>(1);
        });
        doThrow(RuntimeException.class).when(processService).start(any());

        /* When */
        CompletableFuture<Void> future = runAsync(() -> ffmpegDownloader.download());
        Try.run(() -> TimeUnit.MILLISECONDS.sleep(150));
        ffmpegDownloader.pauseDownload();

        /* Then */
        await().atMost(2, TimeUnit.SECONDS).until(() -> {
            assertThat(item).hasStatus(Status.FAILED);
        });
        future.cancel(true);
    }
    
    @Test
    public void should_stop_a_paused_download() throws IOException, URISyntaxException {
        /* Given */
        item.setStatus(Status.PAUSED);
        Process process = mock(Process.class);
        when(ffmpegService.download(anyString(), any(), any())).thenReturn(process);

        /* When */
        CompletableFuture<Void> future = runAsync(() -> ffmpegDownloader.download());
        Try.run(() -> TimeUnit.MILLISECONDS.sleep(50));
        ffmpegDownloader.stopDownload();

        /* Then */
        await().atMost(2, TimeUnit.SECONDS).until(() -> {
            assertThat(item).hasStatus(Status.STOPPED);
        });
        future.cancel(true);
    }

    @Test
    public void should_stop_a_current_download() throws IOException, URISyntaxException {
        /* Given */
        Process process = mock(Process.class);
        when(ffmpegService.download(anyString(), any(), any())).thenReturn(process);

        /* When */
        runAsync(() -> ffmpegDownloader.download());
        Try.run(() -> TimeUnit.MILLISECONDS.sleep(500L));
        ffmpegDownloader.stopDownload();

        /* Then */
        await().atMost(5, TimeUnit.SECONDS).until(() -> {
            assertThat(item).hasStatus(Status.STOPPED);
            verify(process).destroy();
        });
    }
    
    @Test
    public void should_return_target_if_already_set() {
        /* Given */
        ffmpegDownloader.target = Paths.get("/tmp/podcast/file.m3u8");
        /* When */
        Path targetFile = ffmpegDownloader.getTargetFile(Item.DEFAULT_ITEM);
        /* Then */
        assertThat(targetFile).isSameAs(ffmpegDownloader.target);
    }
}

File: backend/src/test/kotlin/com/github/davinkevin/podcastserver/manager/downloader/FfmpegDownloaderTest.kt
package com.github.davinkevin.podcastserver.manager.downloader


import com.github.davinkevin.podcastserver.download.DownloadRepository
import com.github.davinkevin.podcastserver.download.ItemDownloadManager
import com.github.davinkevin.podcastserver.entity.Status
import com.github.davinkevin.podcastserver.entity.Status.*
import com.github.davinkevin.podcastserver.messaging.MessagingTemplate
import com.github.davinkevin.podcastserver.service.ProcessService
import com.github.davinkevin.podcastserver.service.ffmpeg.FfmpegService
import com.github.davinkevin.podcastserver.service.properties.PodcastServerParameters
import com.github.davinkevin.podcastserver.service.storage.FileMetaData
import com.github.davinkevin.podcastserver.service.storage.FileStorageService
import com.github.davinkevin.podcastserver.service.storage.UploadRequest
import net.bramp.ffmpeg.builder.FFmpegBuilder
import net.bramp.ffmpeg.progress.Progress
import net.bramp.ffmpeg.progress.ProgressListener
import org.assertj.core.api.Assertions.assertThat
import org.awaitility.Awaitility.await
import org.junit.jupiter.api.*
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import org.mockito.Mock
import org.mockito.Spy
import org.mockito.invocation.InvocationOnMock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.*
import java.net.URI
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardOpenOption
import java.time.Clock
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.util.*
import java.util.concurrent.TimeUnit.SECONDS
import kotlin.io.path.Path

private val fixedDate = OffsetDateTime.of(2019, 3, 4, 5, 6, 7, 0, ZoneOffset.UTC)

/**
 * Created by kevin on 20/02/2016 for Podcast Server
 */
@ExtendWith(MockitoExtension::class)
class FfmpegDownloaderTest {

    private val item: DownloadingItem = DownloadingItem (
            id = UUID.randomUUID(),
            title = "Title",
            status = Status.NOT_DOWNLOADED,
            url = URI("http://a.fake.url/with/file.mp4?param=1"),
            numberOfFail = 0,
            progression = 0,
            podcast = DownloadingItem.Podcast(
                    id = UUID.randomUUID(),
                    title = "A Fake ffmpeg Podcast"
            ),
            cover = DownloadingItem.Cover(
                    id = UUID.randomUUID(),
                    url = URI("https://bar/foo/cover.jpg")
            )
    )

    @Nested
    inner class DownloaderTest {

        @Mock lateinit var downloadRepository: DownloadRepository
        @Mock lateinit var podcastServerParameters: PodcastServerParameters
        @Mock lateinit var template: MessagingTemplate
        @Mock lateinit var file: FileStorageService

        @Mock lateinit var itemDownloadManager: ItemDownloadManager

        @Mock lateinit var ffmpegService: FfmpegService
        @Mock lateinit var processService: ProcessService
        @Spy val clock: Clock = Clock.fixed(fixedDate.toInstant(), ZoneId.of("UTC"))

        lateinit var downloader: FfmpegDownloader


        @BeforeEach
        fun beforeEach() {
            downloader = FfmpegDownloader(downloadRepository, template, clock, file, ffmpegService, processService)

            downloader
                    .with(DownloadingInformation(item, listOf(item.url, URI.create("http://foo.bar.com/end.mp4")), Path("file.mp4"), "Fake UserAgent"), itemDownloadManager)
        }

        @Nested
        inner class DownloadOperation {

            @BeforeEach
            fun beforeEach() {
                whenever(downloadRepository.updateDownloadItem(any())).thenReturn(0)
            }

            @Test
            fun should_download_file() {
                /* Given */
                whenever(ffmpegService.getDurationOf(any(), any())).thenReturn(500.0)
                whenever(ffmpegService.download(any(), any(), any())).then { i ->
                    writeEmptyFileTo(outputPath(i))
                    (0..100).map { it * 5L }.forEach { sendProgress(i, it)}
                    mock<Process>()
                }
                whenever(processService.waitFor(any())).thenReturn(Result.success(1))
                doAnswer { writeEmptyFileTo(it.getArgument<Path>(0).toString()); null
                }.whenever(ffmpegService).concat(any(), anyVararg())
                doNothing().whenever(file).upload(argThat<UploadRequest.ForItemFromPath> { podcastTitle == item.podcast.title })
                whenever(file.metadata(eq(item.podcast.title), any()))
                    .thenReturn(FileMetaData("video/mp4", 123L))
                whenever(downloadRepository.finishDownload(
                        id = item.id,
                        length = 123L,
                        mimeType = "video/mp4",
                        fileName = Path("file-${item.id}.mp4"),
                        downloadDate = fixedDate
                )).thenReturn(0)

                /* When */
                downloader.run()

                /* Then */
                await().atMost(5, SECONDS).untilAsserted {
                    assertThat(downloader.downloadingInformation.item.status).isEqualTo(FINISH)
                    assertThat(downloader.downloadingInformation.item.progression).isEqualTo(100)
                }
            }

            @Test
            fun `should end on FAILED if a download has failed`() {
                /* Given */
                whenever(ffmpegService.getDurationOf(any(), any())).thenReturn(500.0)

                doAnswer { i -> writeEmptyFileTo(outputPath(i))
                    (0..100).map { it * 5L }.forEach { sendProgress(i, it)}
                    mock<Process>()
                }.whenever(ffmpegService).download(eq(item.url.toASCIIString()), any(), any())

                doAnswer { throw RuntimeException("Error during download of other url") }
                    .whenever(ffmpegService).download(eq("http://foo.bar.com/end.mp4"), any(), any())

                whenever(processService.waitFor(any())).thenReturn(Result.success(1))

                /* When */
                downloader.run()

                /* Then */
                assertThat(downloader.downloadingInformation.item.status).isSameAs(FAILED)
            }

            @Test
            @Disabled("to check that, we need to extract the temp file generation")
            fun `should delete all files if error occurred during download`() {
                /* Given */
                whenever(ffmpegService.getDurationOf(any(), any())).thenReturn(500.0)

                doAnswer { i -> writeEmptyFileTo(outputPath(i))
                    (0..100).map { it * 5L }.forEach { sendProgress(i, it)}
                    mock<Process>()
                }
                        .whenever(ffmpegService).download(eq(item.url.toASCIIString()), any(), any())
                doAnswer { throw RuntimeException("Error during download of other url") }
                        .whenever(ffmpegService).download(eq("http://foo.bar.com/end.mp4"), any(), any())
                whenever(processService.waitFor(any())).thenReturn(Result.success(1))

                /* When */
                downloader.run()

                /* Then */
            }

            @Test
            @Disabled("to check that, we need to extract the temp file generation")
            fun `should delete all files if error occurred during concat`() {
                /* Given */
                whenever(ffmpegService.getDurationOf(any(), any())).thenReturn(500.0)

                doAnswer { i -> writeEmptyFileTo(outputPath(i))
                    (0..100).map { it * 5L }.forEach { sendProgress(i, it)}
                    mock<Process>()
                }
                        .whenever(ffmpegService).download(any(), any(), any())

                whenever(processService.waitFor(any())).thenReturn(Result.success(1))
                whenever(ffmpegService.concat(any(), anyVararg())).then {
                    throw RuntimeException("Error during concat operation")
                }

                /* When */
                downloader.run()

                /* Then */
            }

            private fun outputPath(i: InvocationOnMock) = i.getArgument<FFmpegBuilder>(1).build().last()!!
            private fun sendProgress(i: InvocationOnMock, outTimeMs: Long) =
                    i.getArgument<ProgressListener>(2).progress(Progress().apply { out_time_ns = outTimeMs })
            private fun writeEmptyFileTo(location: String): Path =
                    Files.write(Paths.get(location), "".toByteArray(), StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING)
            private fun numberOfChildrenFiles(location: Path) = Files
                    .newDirectoryStream(location)
                    .map { it }
                    .size
        }

        @Nested
        inner class DuringDownloadOperation {


            @Test
            fun should_stop_a_download() {
                /* Given */
                downloader.downloadingInformation = downloader.downloadingInformation.status(STARTED)
                downloader.process = mock()
                whenever(downloadRepository.updateDownloadItem(any())).thenReturn(0)

                /* When */
                downloader.stopDownload()

                /* Then */
                await().atMost(5, SECONDS).untilAsserted {
                    assertThat(downloader.downloadingInformation.item.status).isEqualTo(STOPPED)
                    verify(downloader.process).destroy()
                }
            }

            @Test
            fun should_failed_to_stop_a_download() {
                /* Given */
                downloader.downloadingInformation = downloader.downloadingInformation.status(STARTED)
                downloader.process = mock()
                doAnswer { throw RuntimeException("Error when executing process") }
                        .whenever(downloader.process).destroy()
                whenever(downloadRepository.updateDownloadItem(any())).thenReturn(0)
                /* When */
                downloader.stopDownload()

                /* Then */
                await().atMost(5, SECONDS).untilAsserted {
                    assertThat(downloader.downloadingInformation.item.status).isEqualTo(FAILED)
                }
            }

        }
    }

    @Nested
    inner class CompatibilityTest {

        @Mock lateinit var downloadRepository: DownloadRepository
        @Mock lateinit var podcastServerParameters: PodcastServerParameters
        @Mock lateinit var template: MessagingTemplate
        @Mock lateinit var file: FileStorageService

        @Mock lateinit var itemDownloadManager: ItemDownloadManager

        @Mock lateinit var ffmpegService: FfmpegService
        @Mock lateinit var processService: ProcessService
        @Spy val clock: Clock = Clock.fixed(fixedDate.toInstant(), ZoneId.of("UTC"))

        lateinit var downloader: FfmpegDownloader

        @BeforeEach
        fun beforeEach() {
            downloader = FfmpegDownloader(
                downloadRepository,
                template,
                clock,
                file,
                ffmpegService,
                processService
            )
        }

        @Test
        fun `should be compatible with multiple urls ending with M3U8 and MP4`() {
            /* Given */
            val di = DownloadingInformation(item, listOf("http://foo.bar.com/end.M3U8", "http://foo.bar.com/end.mp4").map(URI::create), Path("end.mp4"), null)
            /* When */
            val compatibility = downloader.compatibility(di)
            /* Then */
            assertThat(compatibility).isEqualTo(10)
        }

        @Test
        fun `should be compatible with only urls ending with M3U8`() {
            /* Given */
            val di = DownloadingInformation(item, listOf("http://foo.bar.com/end.m3u8", "http://foo.bar.com/end.M3U8").map(URI::create), Path("end.mp4"), null)
            /* When */
            val compatibility = downloader.compatibility(di)
            /* Then */
            assertThat(compatibility).isEqualTo(10)
        }

        @Test
        fun `should be compatible with only urls ending with mp4`() {
            /* Given */
            val di = DownloadingInformation(item, listOf("http://foo.bar.com/end.MP4", "http://foo.bar.com/end.mp4").map(URI::create), Path("end.mp4"), null)
            /* When */
            val compatibility = downloader.compatibility(di)
            /* Then */
            assertThat(compatibility).isEqualTo(10)
        }

        @DisplayName("should be compatible with only one url with extension")
        @ParameterizedTest(name = "{arguments}")
        @ValueSource(strings = ["m3u8", "mp4"])
        fun `should be compatible with only one url with extension`(ext: String) {
            /* Given */
            val di = DownloadingInformation(item, listOf(URI.create("http://foo.bar.com/end.$ext")), Path("end.mp4"), null)
            /* When */
            val compatibility = downloader.compatibility(di)
            /* Then */
            assertThat(compatibility).isEqualTo(10)
        }

        @DisplayName("should not be compatible with")
        @ParameterizedTest(name = "{arguments}")
        @ValueSource(strings = ["http://foo.bar.com/end.webm", "http://foo.bar.com/end.manifest"])
        fun `should not be compatible with`(url: String) {
            /* Given */
            val di = DownloadingInformation(item, listOf(URI.create(url)), Path("end.mp4"), null)
            /* When */
            val compatibility = downloader.compatibility(di)
            /* Then */
            assertThat(compatibility).isEqualTo(Integer.MAX_VALUE)
        }

    }
}
